modules/rostests/apitests/ntdll/RtlCriticalSection.c

```diff
@@ -0,0 +1,374 @@
+Test_Init(void)
+{
+    NTSTATUS Status;
+    ULONG DefaultSpinCount;
+    BOOL HasDebugInfo = (g_Version <= 0x600);
+
+    if ((g_Version >= 0x600) && (g_SysInfo.dwNumberOfProcessors > 1))
+    {
+        DefaultSpinCount = 0x20007d0;
+    }
+    else
+    {
+        DefaultSpinCount = 0;
+    }
+
+    _SEH2_TRY
+    {
+        RtlInitializeCriticalSection(NULL);
+        Status = STATUS_SUCCESS;
+    }
+    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+    {
+        Status = _SEH2_GetExceptionCode();
+    }
+    _SEH2_END;
+    ok_ntstatus(Status, STATUS_ACCESS_VIOLATION);
+
+    Status = RtlInitializeCriticalSection(&CritSect);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_long(CritSect.LockCount, -1);
+    ok_long(CritSect.RecursionCount, 0);
+    ok_ptr(CritSect.OwningThread, NULL);
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+    if (HasDebugInfo)
+    {
+        ok(CritSect.DebugInfo != NULL, "DebugInfo is %p\n", CritSect.DebugInfo);
+        ok(CritSect.DebugInfo != LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+    else
+    {
+        ok(CritSect.DebugInfo == LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+
+    Status = RtlInitializeCriticalSectionAndSpinCount(&CritSect, 0);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_long(CritSect.LockCount, -1);
+    ok_long(CritSect.RecursionCount, 0);
+    ok_ptr(CritSect.OwningThread, NULL);
+    ok_ptr(CritSect.LockSemaphore, NULL);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+    if (HasDebugInfo)
+    {
+        ok(CritSect.DebugInfo != NULL, "DebugInfo is %p\n", CritSect.DebugInfo);
+        ok(CritSect.DebugInfo != LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+    else
+    {
+        ok(CritSect.DebugInfo == LongToPtr(-1), "DebugInfo is %p\n", CritSect.DebugInfo);
+    }
+
+    Status = RtlInitializeCriticalSectionAndSpinCount(&CritSect, 0xFF000000);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+
+    Status = RtlInitializeCriticalSectionAndSpinCount(&CritSect, 0x1234);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_size_t(CritSect.SpinCount, 0x1234);
+
+    if (pfnRtlInitializeCriticalSectionEx != NULL)
+    {
+        _SEH2_TRY
+        {
+            pfnRtlInitializeCriticalSectionEx(NULL, 0, 0);
+            Status = STATUS_SUCCESS;
+        }
+        _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
+        {
+            Status = _SEH2_GetExceptionCode();
+        }
+        _SEH2_END;
+        ok_ntstatus(Status, STATUS_ACCESS_VIOLATION);
+
+        Status = pfnRtlInitializeCriticalSectionEx(&CritSect, 0, 0);
+        ok_ntstatus(Status, STATUS_SUCCESS);
+        ok_long(CritSect.LockCount, -1);
+        ok_long(CritSect.RecursionCount, 0);
+        ok_ptr(CritSect.OwningThread, NULL);
+        ok_ptr(CritSect.LockSemaphore, NULL);
+        ok_size_t(CritSect.SpinCount, DefaultSpinCount);
+        if (HasDebugInfo)
```

Comment:
This branch seems unreachable. `RtlInitializeCriticalSectionEx` only exists on NT 6.x, `HasDebugInfo` is never true for NT 6.x. Is there some kind of compatibility mode setting involved in making this possible?

Reply:
`BOOL HasDebugInfo = (g_Version <= 0x600);`

Smaller or equal :)

Reply:
Ohhh, that dang Vista! :p

Reply:
I assume (but can't recall exactly) that application verifier might be using this field.

And there is this appcompat shim flag:
```
FLAG: AllocDebugInfoForCritSections
  NAME: AllocDebugInfoForCritSections
  FIX_ID: (Binary data)
  FLAG_MASK_KERNEL: 4194304
  GENERAL
```

We don't have those yet, but I could add them?

Reply:
OK, but why?

Reply:
Windows uses it for emulating older systems:
![image](https://github.com/reactos/reactos/assets/610685/64bf86d0-dafa-47c2-a231-ddecd639e8ec)

Reply:
OK, but this has nothing to do with this PR. It would also be easier and less hacky to simply use `if (OsVersion <= _WIN32_WINNT_VISTA)` or something like that.